### PR TITLE
BufferSession Cleanup

### LINKED ISSUE #208: BufferSession is Confusing
`BufferSession` is currently confusing. The `Buffer` field is effectively duplicated across two properties and it is only obvious how to use the class if the source is inspected.

Clean this up and make the API simpler to use.

### PR DESCRIPTION
**Pull request by jcdickinson**
- Clean up BufferSession API.
- Add stream for BufferSession (needs tests).
- Fix some warnings.

Fixes #208 
Introduces #209

### CODE REVIEW

File: src/SourceCode.Clay.Buffers/BufferSession.cs [hunk 1]
```diff
@@ -17,87 +19,85 @@ namespace SourceCode.Clay.Buffers
     /// <seealso cref="System.IDisposable" />
     public struct BufferSession : IEquatable<BufferSession>, IDisposable // MUST be a struct, in order to lessen GC load
     {
+        /// <summary>
+        /// Gets a reference to an empty <see cref="BufferSession"/>.
+        /// </summary>
+        public static BufferSession Empty => new BufferSession(Array.Empty<byte>());
```
Comment (grant-d) [inline src/SourceCode.Clay.Buffers/BufferSession.cs]: `private static ref readonly _empty;`
`... Empty => ref _empty;`

File: src/SourceCode.Clay.Buffers/BufferSession.cs [hunk 2]
```diff
@@ -17,87 +19,85 @@ namespace SourceCode.Clay.Buffers
...
         /// <summary>
-        /// Gets the delineated result (<see cref="BufferSession.Buffer"/> may be overallocated).
+        /// Gets the delineated result.
         /// </summary>
         /// <value>
         /// The result.
         /// </value>
-        public ArraySegment<byte> Result { get; }
+        public ArraySegment<byte> Result => _result.Array == null
+            ? new ArraySegment<byte>(Array.Empty<byte>())
```
Comment (grant-d) [inline src/SourceCode.Clay.Buffers/BufferSession.cs]: `new ArraySegment<byte>(Empty)`

File: src/SourceCode.Clay.Buffers/BufferSession.cs [hunk 3]
```diff
@@ -17,87 +19,85 @@ namespace SourceCode.Clay.Buffers
...
+            return new BufferSession(new ArraySegment<byte>(buffer, 0, minimumLength));
         }
 
-        #endregion
-
-        #region Methods
-
         /// <summary>
-        /// Rents the buffer.
+        /// Wraps a rented <see cref="ArraySegment{T}"/> in a <see cref="BufferSession"/>
```
Comment (grant-d) [inline src/SourceCode.Clay.Buffers/BufferSession.cs]: `/// Wraps an external rented ...`

File: src/SourceCode.Clay.Buffers/BufferSession.cs [hunk 4]
```diff
@@ -112,23 +112,62 @@ public override bool Equals(object obj)
         /// <inheritdoc/>
         public bool Equals(BufferSession other)
```
Comment (grant-d) [inline src/SourceCode.Clay.Buffers/BufferSession.cs]: Turn this into a lambda method

File: src/SourceCode.Clay.Buffers/IO/BufferSessionStream.cs [hunk 5]
```diff
@@ -0,0 +1,283 @@
...
+        public override Task WriteAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+        {
+            Write(buffer, offset, count);
+            return Task.CompletedTask;
+        }
+
+        public override IAsyncResult BeginWrite(byte[] buffer, int offset, int count, AsyncCallback callback, object state)
+        {
+            Write(buffer, offset, count);
+            var async = new SyncAsyncResult(state, count, callback);
```
Comment (grant-d) [inline src/SourceCode.Clay.Buffers/IO/BufferSessionStream.cs]: Prefer not to purposely reuse contextual keywords

File: src/SourceCode.Clay.Buffers/IO/BufferSessionStream.cs [hunk 6]
```diff
@@ -0,0 +1,283 @@
...
+using System.Threading.Tasks;
+
+namespace SourceCode.Clay.IO
+{
+    internal sealed class BufferSessionStream : Stream
+    {
+        #region Fields
+
+        private readonly object _lock;
+        private BufferSession _session;
```
Comment (grant-d) [inline src/SourceCode.Clay.Buffers/IO/BufferSessionStream.cs]: Can this not be `readonly`

Comment (jcdickinson) [reply src/SourceCode.Clay.Buffers/IO/BufferSessionStream.cs]: Nope, if it is the inner `_isDisposed` field is not updated during `Dispose()` and this allows the buffer to be returned to the pool multiple times.

**Pull request merged**